zerver/tests/test_idempotent_request.py

```diff
@@ -0,0 +1,339 @@
+        result_a = self.client_post("/json/messages", post_data, headers=headers)
+        result_b = self.client_post("/json/messages", post_data, headers=headers)
+        content_a = self.assert_json_success(result_a)
+        content_b = self.assert_json_success(result_b)
+        # Again, responses should be identical, indicating duplicate requests are
+        # cached successfully.
+        self.assertEqual(content_a, content_b)
+
+        # Since we changed Idempotency-Key, Only another single idempotency row
+        # is expected.
+        self.assert_length(
+            IdempotentRequest.objects.filter(realm_id=realm.id, user_id=sender.id), 2
+        )
+        # Only another 3rd message should be inserted.
+        self.assert_length(Message.objects.filter(realm_id=realm.id, content="Test message"), 3)
+
+    def test_failed_message_idempotency(self) -> None:
+        """
+        Ensure correct caching of json_error in case of idempotent requests.
+        """
+
+        def get_cached_result_json_content(cached_result: dict[str, Any]) -> dict[str, Any]:
+            return orjson.loads(
+                json_response_from_error(json_error_deserializer(cached_result)).content
+            )
+
+        realm = get_realm("zulip")
+        self.login("hamlet")
+        sender = self.example_user("hamlet")
+        # Set a UUID V4 Idempotency-Key.
+        headers = {"Idempotency-Key": "8e118d41-39d7-4a82-8da9-a6a3162d57eb"}
+
+        post_data = {
+            "type": "channel",
+            "to": "nonexistent_stream",  # this makes it an invalid request,
+            "content": "Test message",
+            "topic": "Test topic",
+        }
+
+        result = self.client_post("/json/messages", post_data, headers=headers)
+        self.assert_json_error(result, "Channel 'nonexistent_stream' does not exist")
+
+        idempotent_requests = IdempotentRequest.objects.filter(realm_id=realm.id, user_id=sender.id)
+
+        # Only one idempotency row should be created.
+        self.assert_length(idempotent_requests, 1)
+
+        # Ensure we mark the row as NOT completed.
+        self.assertEqual(idempotent_requests[0].completed, False)
+
+        # Ensure cached_result is the expected type and not None.
+        self.assertIsInstance(idempotent_requests[0].cached_result, dict)
+
+        # Ensure cached_result has the expected value,
+        # i.e. can deserialized back to the original json error response.
+        self.assertEqual(
+            orjson.loads(result.content),
+            get_cached_result_json_content(
+                cast(dict[str, Any], idempotent_requests[0].cached_result)
+            ),
+        )
+
+        # Send another invalid request but with a different error cause.
+        post_data["to"] = orjson.dumps([99999]).decode()
+        result = self.client_post("/json/messages", post_data, headers=headers)
+        self.assert_json_error(result, "Channel with ID '99999' does not exist")
```

Comment:
If we're posting with the same `Idempotency-Key` header, why don't we get the same cached error response as the previous request?